Migrate a11y article (#3054)
GoogleChrome/web.dev: src/site/content/en/accessible/content-reordering/index.md

---
layout: post

> title: Content re-ordering

Reviewer: ```suggestion
title: Content reordering
```

---
layout: post
title: Content re-ordering
authors:
  - rachelandrew
date: 2020-05-28
description: |
  When creating a layout using CSS, you need to take care that you don't create a disconnected experience for users who are navigation with the keyboard.
tags:
  - blog
  - accessibility
---

The order of content in your document is important for the accessibility of your site.
A screen reader will read out content based on the document order,
using the HTML elements that you have selected to give meaning to that content.
A person navigating the site using a keyboard, rather than a touchscreen or mouse, will tab around the document.
This means that they will jump from active element to active element,
tabbing between links and form fields, once again in the order they exist in the document.

Starting with a well-structured document and using all of the right HTML elements
is therefore a key part of creating an accessible site.
However, it is possible to undo some of that good work when you start using CSS. Let's take a look at why.

## Source vs. visual order

Website navigation is often marked up as a list of links.

> You can then use Flexbox to turn these into a horizontal bar.

Reviewer: ```suggestion
You can then use [Flexbox](/responsive-web-design-basics/#flexbox) to turn these into a horizontal bar.
```

In the Glitch example below, I have created this commonly used pattern.
Click into the example, and tab between the links.
The focus will move in a logical direction from left to right,
the order that we read in English.

<div class="glitch-embed-wrap" style="height: 420px; width: 100%;">
  <iframe
    src="https://glitch.com/embed/#!/embed/flex-nav-source-ordered?path=index.html&previewSize=100"
    title="flex-nav-source-ordered on Glitch"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>
</div>

> If you have created this sort of pattern and then were asked to move "Contact Us",

Reviewer: ```suggestion
If you have created this sort of pattern and then were asked to move **Contact Us**,
```
FYI our style guide convention is to format UI elements in bold: https://developers.google.com/style/ui-elements#formatting

which is second in the source, to the end. You could use the `order` property which works in Flexbox.
Try tabbing through the items in the example below, which has used the `order` property to rearrange the items.

<div class="glitch-embed-wrap" style="height: 420px; width: 100%;">
  <iframe
    src="https://glitch.com/embed/#!/embed/flex-nav-ordered?path=style.css&previewSize=100"
    title="flex-nav-ordered on Glitch"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>

> </div>

Reviewer: So nice having the embedded Glitch here!

The focus jumps to the final item and then back again.
As far as the tab order is concerned that item is the second item.
Visually however, it's the last item.

The above example highlights the problem that we face if we rearrange and reorder content using CSS.
If you were dealing with this issue then the right thing to do would be to change the order in the source,
rather than using CSS.

## Which CSS properties can cause reordering?

Any layout method that allows you to move elements around can cause this problem.

> The following list being key places to consider if you are causing the problem.

Reviewer 2: I found the wording on these two sentences a little hard to grok. I read it a couple times to understand it. Might be worth rephrasing somehow?

Reviewer: ```suggestion
Any layout method that allows you to move elements around can cause this problem.
The following CSS properties commonly cause content reordering problems:
```

- Using `position: absolute` and taking an item out of flow visually.

> - The `order` property in Flexbox and Grid Layout.

Reviewer: ```suggestion
- The `order` property in Flexbox and Grid layout.
```

- The `row-reverse` and `column-reverse` values for `flex-direction` in Flexbox.
- The `dense` value for `grid-auto-flow` in Grid Layout.
- Any positioning by line name or number, or with `grid-template-areas` in Grid Layout.

In this next example, I have created a layout using CSS Grid and positioned the items
using line numbers, without considering where they are in the source.

<div class="glitch-embed-wrap" style="height: 420px; width: 100%;">
  <iframe
    src="https://glitch.com/embed/#!/embed/grid-mixed-up-order?path=index.html&previewSize=100"
    title="grid-mixed-up-order on Glitch"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>
</div>

Try tabbing around this example, and see how the focus jumps about.
This makes for a very confusing experience, especially if this is a long page.

## Testing for the problem

A very simple test is to keyboard navigate through your page. Can you get to everything?
Are there any strange jumps as you do so?

> For a visual demonstration of content reordering,

Reviewer 2: Might want to link the Accessibility Insights text

try the Tab Stop checker in the Accessibility Insights extension for Chrome.
The image below shows the CSS Grid example in that tool.
You can see how the focus has to jump around the layout.

> <figure class="w-figure">

Reviewer: ```suggestion
<figure class="w-figure" style="max-width: 600px">
```

>   <img src="./a11y-insights.jpg" alt="A screenshot of the Accessibility Insights Tool demonstrating the confusing order of items.">

Reviewer: ```suggestion
  <img class="w-screenshot" src="./a11y-insights.jpg" alt="A screenshot of the Accessibility Insights Tool demonstrating the confusing order of items.">
```
